implementation/building.tex

```diff
@@ -0,0 +1,165 @@
+    \item Compiling and ensuring that all dependencies were consistent across builds and systems for the bindings were a tedious task. A lot of the troubles concerning this came back to the point above.
+    \item Linking dependencies with the library required each system to have set up several different \$PATHS to point to the right directories, what worked for one developer on one system might not have worked for different developer on another system.
+    \item The Cargo package manager was developed for exactly these purposes among others.
+\end{itemize}
+
+As already described, Cargo is a tool that provides many operations to build Rust projects that has a certain project structure.
+It is designed to integrate with other existing tools, like GNU Make, which has been important in  building this project.
+When the transition to Cargo started, we focused on structuring the main library and its modules into the directory structure described in \autoref{ssub:project_structure}.
+By invoking the \texttt{cargo build --verbose} command, it was possible to see the output from what Cargo attempted to build when it failed, and then structure the project accordingly.
+
+A big priority was to to shrink the size of the makefiles that was in the project by making them a part of the standard build process for \texttt{emlib} instead, doing this would help us get a long way of ensuring that the builds done by Cargo could be consistent across systems.
+By defining a Rust build script and utilizing a Rust build-dependency called \texttt{gcc} \cite{web:cargo_gcc} (not to be mistaken with the C-compiler with the same name), we were able to compile the C sources from Silabs' emlib and link them with our bindings directly as part of the build process.
+By removing the dependency of manually compiling the C sources, it was easier to start to automatically fetch the other dependencies like the \texttt{core} and \texttt{collections} libraries.
+
+Because the project is for a different processor architecture than the system that it is built on, we had to conditionally cross-compile all the standard Rust libraries that we wanted to utilize for the ARM Cortex-M3.
+We could not utilize the pre-compiled libraries that are already included with \texttt{rustc}, since these only works for the current system architecture.
+We solved this problem by implementing a new Cargo build-dependency, called \texttt{rust-src} \cite{github:rust_src}, whose purpose is to download the entire Rust source code that is compliant with the instance of \texttt{rustc} that is currently compiling the library.
+By making it a task for each build to fetch its own source code, we were guaranteed that the dependencies we used for the project would always compile, independent of the current instance of \texttt{rustc} that was installed on the system.
```

Comment:
Ah, now I got the point you are making. About having to have a local clone of the rust repo that is in sync with the current compiler. This has to be made more explicit.